microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/java/JavaFileTextDocumentService.java

```diff
@@ -152,24 +153,29 @@ public CompletableFuture<Either<List<CompletionItem>, CompletionList>> completio
 				LOGGER.log(Level.SEVERE, "Error while getting java snippet completions", e);
 				return null;
 			}
+			final Integer finalizedCompletionOffset = completionOffset;
 			boolean canSupportMarkdown = true;
 			boolean snippetsSupported = sharedSettings.getCompletionCapabilities().isCompletionSnippetsSupported();
 			CompletionList list1 = new CompletionList();
 			list1.setItems(new ArrayList<>());
-			documents.getSnippetRegistry().getCompletionItems(document, completionOffset, canSupportMarkdown,
-					snippetsSupported, (context, model) -> {
-						if (context != null && context instanceof SnippetContextForJava) {
-							return ((SnippetContextForJava) context).isMatch(projectInfo);
-						}
-						return true;
-					}, projectInfo) //
-					.forEach(item -> {
-						list1.getItems().add(item);
+
+			CompletableFuture<Void> snippetsFuture = microprofileLanguageServer.getLanguageClient().getJavaSnippetCursorContext(javaParams) //
+					.thenAccept((javaCursorContext) -> {
+						documents.getSnippetRegistry().getCompletionItems(document, finalizedCompletionOffset, canSupportMarkdown,

```

Comment:
I wonder if we could avoid using completion offset (finalizedCompletionOffset) and just using the completion position coming from the LSP an dthe completion offset could be resolved on JDT LS side (@jeffmaury is it a problem for you to work with Position (line, char) instead of completion offset on IJ side).

Why I would like to do that? It is because I would like to try to remove the TextDocument for Java file, which maintains offset, although it is just used for completion. We will gain a lot in memory (for big Java file)

Reply:
I think this is an interesting idea, and you should file an issue for it.

The important thing to consider is then we also need some mechanism to synchronize the tab and newline information.

Reply:
Indeed we need those information. If we decide to remove TextDocument, `JavaCursorContextResult` should return those `line information` and if we do that, it should return too the `packageName`.